opc/resource_lbaas_listener.go

```diff
@@ -0,0 +1,304 @@
+				Computed: true,
+			},
+		},
+	}
+}
+
+func resourceListenerCreate(d *schema.ResourceData, meta interface{}) error {
+	lbaasClient, err := meta.(*Client).getLBaaSClient()
+	if err != nil {
+		return err
+	}
+	listenerClient := lbaasClient.ListenerClient()
+
+	var lb lbaas.LoadBalancerContext
+	if loadBalancer, ok := d.GetOk("load_balancer"); ok {
+		lb = getLoadBalancerContextFromID(loadBalancer.(string))
+	}
+
+	input := lbaas.CreateListenerInput{
+		Name:                 d.Get("name").(string),
+		BalancerProtocol:     lbaas.Protocol(d.Get("balancer_protocol").(string)),
+		OriginServerProtocol: lbaas.Protocol(d.Get("server_protocol").(string)),
+		Port:                 d.Get("port").(int),
+	}
+
+	if enabled, ok := d.GetOk("enabled"); ok {
+		input.Disabled = getDisabledStateKeyword(enabled.(bool))
+	}
+
+	if serverPool, ok := d.GetOk("server_pool"); ok {
+		// Only the URI Path is need on Create
+		input.OriginServerPool = getURIRequestPath(serverPool.(string))
+	}
+
+	pathPrefixes := getStringSet(d, "path_prefixes")
+	if len(pathPrefixes) != 0 {
+		input.PathPrefixes = pathPrefixes
+	}
+
+	policies := getStringSet(d, "policies")
+	if len(policies) != 0 {
+		input.Policies = policies
+	}
+
+	sslCerts := getStringSet(d, "certificates")
+	if len(sslCerts) != 0 {
+		input.SSLCerts = sslCerts
+	}
+
+	tags := getStringSet(d, "tags")
+	if len(tags) != 0 {
+		input.Tags = tags
+	}
+
+	virtualHosts := getStringSet(d, "virtual_hosts")
+	if len(virtualHosts) != 0 {
+		input.VirtualHosts = virtualHosts
+	}
+
+	info, err := listenerClient.CreateListener(lb, &input)
+	if err != nil {
+		return fmt.Errorf("Error creating Load Balancer Listener: %s", err)
+	}
+
+	d.SetId(fmt.Sprintf("%s/%s/%s", lb.Region, lb.Name, info.Name))
+	return resourceListenerRead(d, meta)
+}
+
+func resourceListenerRead(d *schema.ResourceData, meta interface{}) error {
+	lbaasClient, err := meta.(*Client).getLBaaSClient()
+	if err != nil {
+		return err
+	}
+	listenerClient := lbaasClient.ListenerClient()
+	name := getLastNameInPath(d.Id())
+	lb := getLoadBalancerContextFromID(d.Id())
+
+	result, err := listenerClient.GetListener(lb, name)
+	if err != nil {
+		// Listener does not exist
+		if client.WasNotFoundError(err) {
+			d.SetId("")
+			return nil
+		}
+		return fmt.Errorf("Error reading Load Balancer Listener %s: %s", d.Id(), err)
+	}
+
+	if result == nil {
+		d.SetId("")
+		return nil
+	}
+
+	d.Set("balancer_protocol", result.BalancerProtocol)
+	d.Set("enabled", getEnabledState(result.Disabled))
+	d.Set("name", result.Name)
+	d.Set("operation_details", result.OperationDetails)
+	d.Set("parent_listener", result.ParentListener)
+	d.Set("port", result.Port)
+	d.Set("server_pool", result.OriginServerPool)
+	d.Set("server_protocol", result.OriginServerProtocol)
+	d.Set("state", result.State)
+	d.Set("uri", result.URI)
+	d.Set("load_balancer", fmt.Sprintf("%s/%s", lb.Region, lb.Name))
+
+	if err := setStringList(d, "path_prefixes", result.PathPrefixes); err != nil {
```

Comment:
I wonder if we want to do a `setStringSet` instead of `setStringList`. Let's add a test that uses some of the `TypeSet` attributes and see if they get set correctly if we use `setStringList`

Reply:
@mbfrahry I'll change it to use just `d.Set()`, theres not need for a separate `setStringSet()`.  Is there a existing method to test for set equality for the acceptance tests? The `resource.TestCheckResourceAttr()` only compares individual string values.